### ISSUE TITLE
Allow use of #define __HIP_ARCH__ to be set to GPU generation like Nvidia (for ex. 900 for Vega)

### ISSUE THREAD

**Issue by oscarbg** [ROCm-Developer-Tools/HIP#2915]
Hi,
first don't know if this is the correct place to ask for this issues but anyway will ask here..
in https://rocmdocs.amd.com/en/latest/Programming_Guides/HIP-porting-guide.html
it's said that HIP provides a portable mechanism to port code using __CUDA_ARCH__ like that:
```
//#if (__CUDA_ARCH__ >= 130)   // non-portable
if __HIP_ARCH_HAS_DOUBLES__ {  // portable HIP feature query
```

it's good but still we should have a mechanism to check inside code which AMD architecture the code is being compiled and make use of advanced features of this architecture.. for example imagine I'm compiling for Vega I know has fast half2 operations (rapid packed math)  with so I use --offload-arch=gfx900 but in code I should have a fallback code for Polaris --offload-arch=gfx801 so I would like to have code like this..

```
__device__ __forceinline__ half2 computationsum(half2 a, half2 b, half2 c){
#if __HIP_ARCH__ >=900
        return __hfma2(a,b,c);
#else
	half2 temp;
	temp.x=__float2half(__half2float(a.x)+__half2float(b.x)+__half2float(c.x));
	temp.y=__float2half(__half2float(a.y)+__half2float(b.y)+__half2float(c.y));
	return temp;
	
	//return half2(__half2float(a.x)+__half2float(b.x)+__half2float(c.x),__half2float(a.y)+__half2float(b.y)+__half2float(c.y));
#endif
}

__device__ __forceinline__ half2 computationdiv(half2 a, half2 b){
#if __HIP_ARCH__ >=900
        return __h2div(a,b);
#else
	half2 temp;
	temp.x=__float2half(__half2float(a.x)/__half2float(b.x));
	temp.y=__float2half(__half2float(a.y)/__half2float(b.y));
	return temp;
#endif        

}

```

also for upcoming RDNA3 gfx11 (#define __HIP_ARCH__ 1100)  I want to use sudot8 new instruction..
how I can check from code what architecture I'm compiling code for?

```
  __global__ void partition(int *res) {
      int laneId = threadIdx.x & 0x1f;
      int value;
      value = valores[laneId]; 
	  int mask;
	  
  
  #if __HIP_ARCH__ >= 1100
  //RDNA3 code
	  int A = mask; int B = value; int C = laneId;
	  mask = __builtin_amdgcn_sudot8(false, A, true, B, C, false);
  #else
    //unoptimized code
  #endif
  
  }
```
and can't enable __builtin_amdgcn_sudot8 unconditionally on all arches as there aren't fallbacks:
If I compile this code for gfx1030 I would get  from clang:
```
test.hip:62:9: error: '__builtin_amdgcn_sudot8' needs target feature dot8-insts
        mask = __builtin_amdgcn_sudot8(false, A, true, B, C, false);
               ^
35 warnings and 1 error generated when compiling for gfx1030.
```
hope it motivates exposing it as even you can even keep adding things like __HIP_ARCH_HAS_DOT8_INSTS_  this list will grow faster than checking for architecture..
thanks..

**Comment by yxsamliu** [comment]
we do not have `__HIP_ARCH__`  because gfx arch is not necessarily in linear order, i.e. a feature available at a lower number may not be available at a higher number.

To determine the availability of clang builtin function you can use __has_builtin.

There are also pre-defined macros which can help, e.g. `__gfx906__`, `__gfx1030__`, `__GFX9__`, `__GFX10__`

**Comment by oscarbg** [comment]
thanks @yxsamliu.. interesting..
I assume because of differentation of MI100 MI200 having new features and being Vega based so gfx9xx and consumer now RDNA,RDNA2 gfx10xx but not having some of the MIxxx features.. 

so you are right, but at least on Radeon consumer cards I expect linear order as new desktop cards have all previous Radeon features, right? can you comment on this?

also thanks for the hints on  __has_builtin, and macros __gfx906__,etc..
so basically for my gfx11 example will use __GFX11__ to check in code for dot8-insts..

**Comment by yxsamliu** [comment]
some builtin functions may not be supported at higher gfx even though it is supported at lower gfx. e.g. `__builtin_amdgcn_s_memtime` is not in gfx11* but is in gfx10*. The reason is that its function is realized by other instructions due to ISA change.

**Comment by oscarbg** [comment]
ok thanks.. great to know..
